### PR TITLE
Add prefetching to SiteMediaPageViewController and coalescing to ImageDownloader

### PR DESCRIPTION
**Pull request by kean**
- Add prefetching to `SiteMediaPageViewController`.
- Add coalescing to `ImageDownloader` that powers the prefetching. This change is also a prerequisite to removing `WPImageSource` because the improved downloader now has the same (and more) of its features.

To test:

- Open "Media"
- Tap on one of the items to open the item details
- Scroll left or right (test both direction)
- Verify that if you scroll slow enough, the images are prefetching in the direction you are scrolling

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/107fb4e9-4ede-4ea6-b15e-dee02a15a724

This is going to be tricky to test (maybe with a proxy?), but if, when you scroll to the side, the prefetching task is still pending, there should be no new request sent thanks to the coalescing support in `ImageDownloader.`

## Regression Notes
1. Potential unintended areas of impact: Site Media
2. What I did to test those areas of impact (or what existing automated tests I relied on): yes
3. What automated tests I added (or what prevented me from doing so): yes

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Utility/Media/ImageDownloader.swift [hunk 1]
```diff
@@ -103,6 +105,35 @@ actor ImageDownloader {
...
+        })
+        let subscriptionID = UUID()
+        task.subscriptions.insert(subscriptionID)
+        tasks[requestKey] = task
+
+        return try await withTaskCancellationHandler {
+            try await task.task.value
+        } onCancel: {
+            Task {
+                await self.unsubscribe(subscriptionID, key: requestKey)
```
Comment (momo-ozawa) [inline WordPress/Classes/Utility/Media/ImageDownloader.swift:120]: [Q] Why is unsubscribe(_: key:) async without explicitly being marked as async?

Comment (momo-ozawa) [reply WordPress/Classes/Utility/Media/ImageDownloader.swift:120]: Oh - is it because Task { ... } isn't isolated to ImageDownloader?

Comment (kean) [reply WordPress/Classes/Utility/Media/ImageDownloader.swift:120]: The `onCancel` closure is not isolated to `ImageDownloader`, but the `unsubscribe` needs to be actor-isolated on `ImageDownloader`. The only way to call an actor-isolated method from "outside" is from an async function. It's a bit more complicated than I maybe wanted it to be, but it's still better than the traditional approach of wrapping every method in `queue.async` / GCD.

**Review by momo-ozawa — APPROVED** [review]
Works as described ✨

**Pull request merged**